File: 399-EvaluateDivision/399-EvaluateDivision.py
# Last updated: 4/29/2026, 10:30:37 AM
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        from collections import defaultdict
        pathMap = defaultdict(set)
        valueMap = defaultdict(lambda: None)
        for idx, item in enumerate(equations):
            u, v = item
            pathMap[u].add(v)
            pathMap[v].add(u)
            key1 = str(u) + "-" + str(v)
            key2 = str(v) + "-" + str(u)
            valueMap[key1] = values[idx]
            valueMap[key2] = 1.0 / values[idx]
        print(pathMap)
        print(valueMap)
        
        def dfs(u, v, visited, val):
            if len(pathMap[u]) == 0:
                return -1.0
            if u == v:
                return 1.0
            visited[u] = True
            pathKey = u + "-" + v
            if valueMap[pathKey] != None:
                return val * valueMap[pathKey]
            for end in pathMap[u]:
                if not visited[end]:
                    key2 = u + "-" + end
                    newVal = val * valueMap[key2]
                    ret = dfs(end, v, visited, newVal)
                    if ret != -1.0:
                        return ret
            return -1.0
        res = []
        for u, v in queries:
            visited = defaultdict(lambda: False)
            val = dfs(u, v, visited, 1.0)
            res.append(val)
        return res

File: 399-EvaluateDivision/399-EvaluateDivision.js
// Last updated: 4/29/2026, 10:30:32 AM
/**
 * @param {string[][]} equations
 * @param {number[]} values
 * @param {string[][]} queries
 * @return {number[]}
 */
var calcEquation = function(equations, values, queries) {
    const map = {};
    const n = equations.length;
    for(let i = 0; i < n; i++) {
        const [u, v] = equations[i];
        const w = values[i];
        if(map[u] === undefined) {
            map[u] = {};
        }
        map[u][v] = w;
        if(map[v] === undefined) {
            map[v] = {};
        }
        map[v][u] = 1 / w;
    }
    console.log(map);

    const dfs = (start, end, val, visitedMap) => {
        console.log(`start: ${start}, end: ${end}, val: ${val}`)
        if(map[start] !== undefined && start === end) {
            console.log(`return ${val}`)
            return val;
        }
        visitedMap[start] = true;
        const mapValue = map[start];
        console.log(`mapValue:`)
        console.log(mapValue)
        if(mapValue !== undefined) {
            const keys = Object.keys(mapValue);
            for(const key of keys) {
                const w = mapValue[key];
                console.log(`key: ${key}, w: ${w}`)
                if(!visitedMap[key]) {
                    const res = dfs(key, end, val * w, visitedMap);
                    if(res !== undefined) {
                        return res;
                    }
                }
            }
        }
    }

    const output = [];
    for(const query of queries) {
        const [u, v] = query;
        const visitedMap = {};
        const res = dfs(u, v, 1, visitedMap);
        if(res !== undefined) {
            output.push(res);
        } else {
            output.push(-1);
        }
    }
    return output;
};